Spurious widening precision loss at `EbpfDomain::operator&` bottom branch

PREVAIL rejects the concretely-safe program `ebpf-samples/linux-selftests/loop3.o` section `raw_tracepoint/consume_skb` function `while_true` with `Invalid type (r3.type in {number, ctx, stack, packet, shared})`, even though the concrete execution keeps `r3.type` as `number` throughout the loop. The rejection is a precision artifact of widening, traceable to a single line in `EbpfDomain::operator&` that discards the post-meet state on the bottom branch and, together with `SplitDBM::widen` not short-circuiting on bottom inputs, causes correlations to be dropped across loop iterations.

## Reproducer

At `472a486b` (v0.2.1):

```sh
bin/prevail --failure-slice \
    ebpf-samples/linux-selftests/loop3.o \
    raw_tracepoint/consume_skb while_true
```

Output:

```
[ERROR] Invalid type (r3.type in {number, ctx, stack, packet, shared})
[LOCATION] 6
```

Same result with `-v` and with no flag (prints `FAIL:
raw_tracepoint/consume_skb/while_true`).

## The program

Disassembly:

```
0:  r2 = 0
1:  *(u64 *)(r10 - 8) = r2         ; counter := 0
2:  *(u64 *)(r10 - 16) = r2        ; acc := 0
3:  r2 = 4294967296 ll             ; r2 := 2^32  (loop bound)
5:  r3 = *(u64 *)(r10 - 8)         ; counter into r3
6:  r3 += 1
7:  *(u64 *)(r10 - 8) = r3         ; counter := counter + 1
8:  r3 = *(u64 *)(r1 + 80)         ; r3 := ctx->something
9:  r4 = *(u64 *)(r10 - 16)
10: r4 += r3                       ; acc += ctx->something
11: *(u64 *)(r10 - 16) = r4
12: r3 = *(u64 *)(r10 - 8)         ; reload counter
13: if r3 < r2 goto -9 <5>         ; loop while counter < 2^32
14: r0 = *(u64 *)(r10 - 16)
15: exit
```

`r3` is always numeric: the counter slot `s[-8]` is initialised to
zero numeric at label 1, incremented in place, and the ctx-load at
label 8 is already numeric by typing. The assertion
`r3.type in {number, ctx, stack, packet, shared}` at label 6 cannot
fail concretely.

## Where precision is lost

The failure trace's widened loop-head pre-invariant is:

```
s[4080...4087].svalue=0, s[4080...4087].uvalue=0,
s[4088...4095].svalue=0, s[4088...4095].uvalue=0
```

Concrete values, no correlations. In particular, the relation
between `s[4088..4095]` and `r3` — needed to conclude that the
value loaded from `s[-8]` at label 5 is a number — is gone.

Widening *can* preserve `s[4088...4095].svalue = r3.svalue` (and the
corresponding range `s[4088...4095].svalue ∈ [0, r2.svalue)`) across
iterations: both sides of the widening refer to the same variables
with the same edges. With that correlation, the loaded `r3` at
label 5 carries `r3.svalue = s[4088...4095].svalue ∈ [0, 2^32)`,
`r3.uvalue = r3.svalue`, hence `r3.type = number`, and the
assertion at label 6 trivially holds.

## Root cause

`EbpfDomain::operator&` in `src/crab/ebpf_domain.cpp`:

```cpp
EbpfDomain EbpfDomain::operator&(const EbpfDomain& other) const {
    auto res = state & other.state;
    if (!res.is_bottom()) {
        return {std::move(res), stack & other.stack};
    }
    return bottom();   // ← discards `res` even though `res` already is bottom
}
```

And `EbpfDomain::widen`:

```cpp
EbpfDomain EbpfDomain::widen(const EbpfDomain& other, const bool to_constants) const {
    EbpfDomain res{this->state.widen(other.state), stack.widen(other.stack)};
    if (to_constants) {
        return res & get_constant_limits();  // ← this meet can go bottom
    }
    return res;
}
```

On the first widening iteration (`to_constants = true`), the widened
state `res` may contain abstract values that violate the constant
limits (e.g. `r3.uvalue` widened to `+∞`, clamped to
`[0, u32::max()]`; the concrete `r3` reaches `r2 = 2^32` which just
exceeds `u32::max()`). The `&` meet then produces bottom.

The operator discards the post-meet state by returning `bottom()` — a
freshly constructed `TypeToNumDomain` with `set_to_bottom()` — which
has no registered variables and an empty zone-domain graph. The
correlations carried by `res` (including
`s[4088...4095].svalue = r3.svalue`) are lost.

If `res & get_constant_limits()` instead returned the post-meet
state directly — still `is_bottom() == true`, but with the residual
variable set and graph edges from `res` — the correlations are
preserved in the representation, and subsequent `widen` calls (which
do not short-circuit on bottom inputs in `SplitDBM::widen`) propagate
the correlation into the next iteration.

## Proposed fix

Either:

1. **`EbpfDomain::operator&` preserves residual state on bottom**
   (strictly more precise; empirically makes `loop3.o` verify):

   ```cpp
   EbpfDomain EbpfDomain::operator&(const EbpfDomain& other) const {
       auto res = state & other.state;
       if (!res.is_bottom()) {
           return {std::move(res), stack & other.stack};
       }
       // res is bottom; keep it (with residual vertex/edge structure)
       // rather than discarding via bottom(), so downstream widen can
       // still read the shape of the combined state.
       return {std::move(res), ArrayDomain{}};
   }
   ```

   Both forms are sound (`bottom ⊆ anything`); the second is strictly
   more precise in conjunction with the current `SplitDBM::widen`
   (which does not short-circuit bottom inputs).

2. **Short-circuit `SplitDBM::widen` on bottom inputs**, so that
   `widen(bottom, X) = X` regardless of residual structure in
   `bottom`, restoring the theoretical expectation. Then
   `EbpfDomain::operator&`'s behaviour on the bottom branch is
   immaterial.

(1) is the minimal change that fixes `loop3.o`. (2) is the
principled fix — it makes widening independent of the representation
of bottom, which is a reasonable invariant for an abstract domain
that nominally has a unique bottom element.

## How it was found

Discovered during work on the Rust port
(https://github.com/elazarg/prevail-rust). A mechanical refactor
there inadvertently changed `EbpfDomain::meet` to keep the post-meet
state on the bottom branch (equivalent to fix (1) above) and
`loop3.o` flipped from rejected to accepted. Bisecting the refactor
isolated the single line. The C++ verifier and the (now-fixed) Rust
port share the same behaviour and the same rejection; the precision
improvement is available to both with either proposed fix.

## Other tests

No regressions in the YAML / ELF verify / conformance suites when
applying fix (1) on the Rust side; all 969 sample parity cases
agree. `loop3.o` moves from FAIL to PASS; no other case changes
outcome.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,13 @@
 # Changelog
 
+## v0.2.2 (2026-04-15)
+
+### Bug fixes
+
+- Fix spurious widening precision loss in `EbpfDomain` that rejected
+  concretely-safe programs using numeric counters stored on the stack
+  across loops (#1071). Regression from #916.
+
 ## v0.2.1 (2026-04-14)
 
 Bug fixes, API enhancements, and infrastructure updates.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -9,7 +9,7 @@ if (NOT DEFINED CMAKE_MSVC_RUNTIME_LIBRARY)
   set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>DLL")
 endif ()
 
-project(prevail VERSION 0.2.1)
+project(prevail VERSION 0.2.2)
 
 # NOTE: Do not use CMAKE_SOURCE_DIR in this project.
 # When prevail is consumed via add_subdirectory()/FetchContent, CMAKE_SOURCE_DIR refers to

--- a/examples/using_installed_package/CMakeLists.txt
+++ b/examples/using_installed_package/CMakeLists.txt
@@ -9,7 +9,7 @@ set(CMAKE_CXX_STANDARD 20)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 # This will fail if prevail wasn't installed correctly
-find_package(prevail 0.2.1 REQUIRED)
+find_package(prevail 0.2.2 REQUIRED)
 
 add_executable(install_test main.cpp)
 target_link_libraries(install_test PRIVATE prevail::prevail)

--- a/examples/using_installed_package/README.md
+++ b/examples/using_installed_package/README.md
@@ -44,7 +44,7 @@ build\Release\install_test.exe
 ### CMakeLists.txt
 
 ```cmake
-find_package(prevail 0.2.1 REQUIRED)
+find_package(prevail 0.2.2 REQUIRED)
 
 add_executable(myapp main.cpp)
 target_link_libraries(myapp PRIVATE prevail::prevail)

--- a/examples/using_installed_package/TESTING_DETAILS.md
+++ b/examples/using_installed_package/TESTING_DETAILS.md
@@ -13,7 +13,7 @@ installation mechanism (not API stability).
     - Proper permissions are set
 
 2. **CMake Package Discovery**
-    - `find_package(prevail 0.2.1 REQUIRED)` succeeds
+    - `find_package(prevail 0.2.2 REQUIRED)` succeeds
     - Version checking works correctly
     - Package config files are valid
 

--- a/src/crab/ebpf_domain.cpp
+++ b/src/crab/ebpf_domain.cpp
@@ -180,6 +180,12 @@ static const EbpfDomain& get_constant_limits() { return constant_limits.get(); }
 void EbpfDomain::clear_thread_local_state() { constant_limits.clear(); }
 
 EbpfDomain EbpfDomain::widen(const EbpfDomain& other, const bool to_constants) const {
+    if (is_bottom()) {
+        return other;
+    }
+    if (other.is_bottom()) {
+        return *this;
+    }
     EbpfDomain res{this->state.widen(other.state), stack.widen(other.stack)};
 
     if (to_constants) {

--- a/src/test/test_verify_linux_selftests.cpp
+++ b/src/test/test_verify_linux_selftests.cpp
@@ -71,9 +71,7 @@ TEST_PROGRAM_FAIL("linux-selftests", "bloom_filter_map.o", "fentry/__x64_sys_get
 // register type refinement is too imprecise in this control-flow pattern
 TEST_SECTION_FAIL("linux-selftests", "freplace_get_constant.o", "freplace/get_constant",
                   verify_test::VerifyIssueKind::VerifierTypeTracking)
-// Type precision lost through loop join: r3 loaded from stack loses type after widening
-TEST_SECTION_FAIL("linux-selftests", "loop3.o", "raw_tracepoint/consume_skb",
-                  verify_test::VerifyIssueKind::VerifierTypeTracking)
+TEST_SECTION("linux-selftests", "loop3.o", "raw_tracepoint/consume_skb")
 // register type refinement is too imprecise in this control-flow pattern
 TEST_PROGRAM_FAIL("linux-selftests", "map_ptr_kern.o", ".text", "check", 19,
                   verify_test::VerifyIssueKind::VerifierTypeTracking)

--- a/test-data/elf_inventory.json
+++ b/test-data/elf_inventory.json
@@ -8230,15 +8230,6 @@
                 "invalid": false
               }
             ]
-          },
-          "test_overrides": {
-            "sections": {
-              "raw_tracepoint/consume_skb": {
-                "kind": "VerifierTypeTracking",
-                "reason": "Type precision lost through loop join: r3 loaded from stack loses type after widening",
-                "status": "expected_failure"
-              }
-            }
           }
         },
         "loop4.o": {

--- a/test-data/loop.yaml
+++ b/test-data/loop.yaml
@@ -558,3 +558,51 @@ post:
   - pc[1]=r0.svalue+1
 
 messages: []
+
+---
+test-case: stack counter type preserved through widening (issue 1071)
+
+pre: [
+  "r1.type=ctx", "r1.ctx_offset=0", "r1.svalue=[1, 2147418112]",
+  "r2.type=number", "r2.svalue=4294967296", "r2.uvalue=4294967296",
+  "r10.type=stack", "r10.stack_offset=4096", "r10.svalue=[4096, 2147418112]",
+  "meta_offset=0",
+  "packet_size=0",
+]
+
+code:
+  <init>: |
+    r3 = 0
+    *(u64 *)(r10 - 8) = r3
+  <loop>: |
+    r3 = *(u64 *)(r10 - 8)
+    r3 += 1
+    *(u64 *)(r10 - 8) = r3
+    if r3 < r2 goto <loop>
+  <exit>: |
+    r0 = *(u64 *)(r10 - 8)
+    exit
+
+post:
+  - r0.type=number
+  - r0.svalue=4294967296
+  - r0.uvalue=4294967296
+  - r2.type=number
+  - r2.svalue=4294967296
+  - r2.uvalue=r2.svalue
+  - r3.type=number
+  - r3.svalue=r2.svalue
+  - r3.uvalue=r2.svalue
+  - r10.type=stack
+  - r10.stack_offset=4096
+  - r10.svalue=[4096, 2147418112]
+  - r1.type=ctx
+  - r1.ctx_offset=0
+  - r1.svalue=[1, 2147418112]
+  - s[4088...4095].type=number
+  - s[4088...4095].svalue=r2.svalue
+  - s[4088...4095].uvalue=r2.svalue
+  - meta_offset=0
+  - packet_size=0
+
+messages: []

